File: Java/139-WordBreak.java
public class Solution {
    public boolean wordBreak(String s, Set<String> wordDict) {
        boolean[] breakable = new boolean[s.length()+1];
        breakable[0] = true;
        
        for (int i = 1; i <= s.length(); i++){
            for (int j = 0; j < i; j++){
                if (breakable[j] && wordDict.contains(s.substring(j, i))){
                    breakable[i] = true;
                    break;
                }
            }
        }
        return breakable[s.length()];
    }
}

public class Solution {
    public boolean wordBreak(String s, Set<String> wordDict) {
        if (s.length() == 0 || wordDict == null || wordDict.size() == 0){
            return false;
        }
        boolean[] breakable = new boolean[s.length()+1];
        breakable[0] = true;
        for (int i = 1; i <= s.length(); i++){
            for (int j = 0; j <= i && !breakable[i]; j++){
                breakable[i] = wordDict.contains(s.substring(j, i)) && breakable[j];
            }
        }
        return breakable[s.length()];
    }
}

File: cpp/139-WordBreak.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp(s.size() + 1, false);
        dp[0] = true;
        unordered_set<string> wordSet(wordDict.begin(), wordDict.end());
        //i: current string;
        //j: breaking point of current string
        //dp[0]: if string is null
        for (int i = 1; i <= s.size(); i++) {
            for (int j = 0; j < i; j++) {
                if (wordSet.find(s.substr(j, i - j)) != wordSet.end() && dp[j]) {
                    //cout << s.substr(i, j) << endl;
                    dp[i] = true;
                    if (dp[s.size()]) {
                        return true;
                    }
                }
            }
        }
        return dp[s.size()];
    }
};
